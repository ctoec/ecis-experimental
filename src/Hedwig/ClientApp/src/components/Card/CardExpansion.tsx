import React, { Children, useState, isValidElement } from 'react';
import cx from 'classnames';

type CardExpansionProps = {};

export const CardExpansion: React.FC<CardExpansionProps> = ({ children }) => {
	return <>{children}</>;
};
