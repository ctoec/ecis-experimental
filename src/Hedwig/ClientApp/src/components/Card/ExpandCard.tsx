import React, { useContext, isValidElement, ReactEventHandler, Children } from 'react';
import { CardContext } from '..';

/**
 * Component for declaratively expanding/collapsing card
 * @param props
 */
export const ExpandCard: React.FC = ({ children }) => {
	const { toggleExpanded } = useContext(CardContext);

	if (!isValidElement(children)) {
		throw new Error('Invalid children to ExpandCard');
	}

	return (
		<>
			{Children.map(children, child => {
				const { type: Type, props, key } = child;
				return (
					<Type
						{...key}
						{...props}
						onClick={(e: ReactEventHandler) => {
							toggleExpanded();
							if (props.onClick) {
								props.onClick(e);
							}
						}}
					/>
				);
			})}
		</>
	);
};
