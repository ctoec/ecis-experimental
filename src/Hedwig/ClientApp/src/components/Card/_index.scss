.oec-card {
	width: 100%;

	border-width: 3px;
	@include u-border('gray-10');

	&--secondary {
		@include u-bg('gray-5');
	}

	.oec-card-tag {
		position: relative;
		float: left;
		top: -3px;
		left: -3px;
	}

	.oec-card-cell {
		@include u-padding-y(4);
		@include u-padding-x(6);
	}

	.oec-card-divider + .oec-card-cell {
		@include u-bg('white');
	}

	.oec-card-divider {
		border-width: 3px;
		@include u-border-bottom('gray-10');
	}
}
