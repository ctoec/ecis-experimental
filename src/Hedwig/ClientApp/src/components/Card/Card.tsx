import React, {
	Children,
	useEffect,
	useState,
	isValidElement,
	PropsWithChildren,
	createContext,
} from 'react';
import cx from 'classnames';
import { Tag } from '..';
import { CardExpansion } from './CardExpansion';

type CardContextType = {
	isExpanded: boolean;
	toggleExpanded: () => void;
};

const CardContext = createContext<CardContextType>({
	isExpanded: false,
	toggleExpanded: () => {},
});

const { Provider: CardProvider } = CardContext;
export { CardContext };

export type CardProps = {
	appearance?: 'primary' | 'secondary';
	className?: string;
	expanded?: boolean;
	onExpansionChange?: (_: boolean) => void;
	forceClose?: boolean;
	showTag?: boolean;
};

export function Card({
	appearance,
	className,
	expanded = false,
	onExpansionChange,
	forceClose,
	showTag = false,
	children,
}: PropsWithChildren<CardProps>) {
	const [previousIsExpanded, setPreviousIsExpanded] = useState<boolean>();
	const [isExpanded, setIsExpanded] = useState(expanded);
	const updateExpanded = (_: boolean) => {
		setPreviousIsExpanded(isExpanded);
		setIsExpanded(_);
	};
	const toggleExpanded = () => updateExpanded(!isExpanded);

	useEffect(() => {
		if (onExpansionChange && previousIsExpanded !== undefined) {
			onExpansionChange(isExpanded);
		}
	}, [isExpanded]);

	useEffect(() => {
		if (forceClose) {
			updateExpanded(false);
		}
	}, [forceClose]);

	return (
		<CardProvider
			value={{
				isExpanded,
				toggleExpanded,
			}}
		>
			<div
				className={cx(
					'oec-card',
					{
						[`oec-card--${appearance}`]: appearance,
					},
					className
				)}
			>
				{showTag && <Tag className="oec-card-tag" text="NEW" color="theme-color-primary" />}
				<div className="oec-card-cell">
					{Children.map(children, child => {
						if (!isValidElement(child)) {
							throw new Error('Invalid card child element');
						}
						const type = (typeof child.type === 'string') ? child.type : child.type.name;
						if (type !== CardExpansion.name) {
							return child;
						}
					})}
				</div>
				<div
					className={cx({
						'oec-card-divider': isExpanded,
					})}
				></div>
				<div className={cx('oec-card-cell', 'oec-card-expansion')} hidden={!isExpanded}>
					{Children.map(children, child => {
						if (!isValidElement(child)) {
							throw new Error('Invalid card child element');
						}
						const type = (typeof child.type === 'string') ? child.type : child.type.name;
						if (type === CardExpansion.name) {
							return child;
						}
					})}
				</div>
			</div>
		</CardProvider>
	);
}
