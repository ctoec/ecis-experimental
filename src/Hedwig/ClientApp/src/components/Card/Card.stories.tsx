import React from 'react';
import { storiesOf } from '@storybook/react';
import { Card } from '..';
import { CardExpansion } from './CardExpansion';
import { ExpandCard } from './ExpandCard';
import { Button } from '../Button/Button';

storiesOf('Card', module)
	.add('Primary with Text', () => {
		return (
			<Card>
				<span>I am a card!</span>
			</Card>
		);
	})
	.add('Secondary with Text', () => {
		return (
			<Card appearance="secondary">
				<span>I am a card!</span>
			</Card>
		);
	})
	.add('Compact Primary with Text', () => {
		return (
			<Card className="width-max-content important">
				<span>I am a compact card!</span>
			</Card>
		);
	})
	.add('Primary with Grid', () => {
		return (
			<Card>
				<div className="usa-grid">
					<div className="grid-row text-bold">
						<div className="grid-col">First name</div>
						<div className="grid-col">Last name</div>
					</div>
					<div className="grid-row">
						<div className="grid-col">Daniel</div>
						<div className="grid-col">Radcliff</div>
					</div>
				</div>
			</Card>
		);
	})
	.add('Secondary with Grid', () => {
		return (
			<Card appearance="secondary">
				<div className="usa-grid">
					<div className="grid-row text-bold">
						<div className="grid-col">First name</div>
						<div className="grid-col">Last name</div>
					</div>
					<div className="grid-row">
						<div className="grid-col">Daniel</div>
						<div className="grid-col">Radcliff</div>
					</div>
				</div>
			</Card>
		);
	})
	.add('With closed card expansion', () => {
		return (
			<Card>
				<div className="usa-grid">
					<div className="grid-row text-bold">
						<div className="grid-col">First name</div>
						<div className="grid-col">Last name</div>
					</div>
					<div className="grid-row">
						<div className="grid-col">Daniel</div>
						<div className="grid-col">Radcliff</div>
					</div>
				</div>
				<CardExpansion>
					<div>More information here but you can't see :'(</div>
				</CardExpansion>
			</Card>
		);
	})
	.add('With open card expansion', () => {
		return (
			<Card expanded={true}>
				<div className="usa-grid">
					<div className="grid-row text-bold">
						<div className="grid-col">First name</div>
						<div className="grid-col">Last name</div>
					</div>
					<div className="grid-row">
						<div className="grid-col">Daniel</div>
						<div className="grid-col">Radcliff</div>
					</div>
				</div>
				<CardExpansion>
					<div>You can see me!! :D</div>
				</CardExpansion>
			</Card>
		);
	})
	.add('With card expansion and card expander', () => {
		return (
			<Card expanded={true}>
				<div className="usa-grid">
					<div className="grid-row">
						<div className="grid-col">
							<div className="grid-row text-bold">
								<div className="grid-col">First name</div>
								<div className="grid-col">Last name</div>
							</div>
							<div className="grid-row">
								<div className="grid-col">Daniel</div>
								<div className="grid-col">Radcliff</div>
							</div>
						</div>
						<div className="grid-col">
							<ExpandCard>
								<Button appearance="unstyled" text="Click on me to toggle showing the expansion" />
							</ExpandCard>
						</div>
					</div>
				</div>
				<CardExpansion>
					<div>You can see me!! :D</div>
				</CardExpansion>
			</Card>
		);
	});
