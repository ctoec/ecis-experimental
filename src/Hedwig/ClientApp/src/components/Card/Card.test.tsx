import React from 'react';
import { render, act } from '@testing-library/react';
import { Card } from './Card';
import { CardExpansion } from './CardExpansion';
import { ExpandCard } from './ExpandCard';

it('matches snapshot', () => {
	const { asFragment } = render(
		<Card>
			<p>Hi there</p>
		</Card>
	);
	expect(asFragment()).toMatchSnapshot();
});

describe('card with card expansion', () => {
	it('hides expansion when expanded is false', () => {
		const { getByText } = render(
			<Card expanded={false}>
				<p>Shown</p>
				<CardExpansion>Hidden</CardExpansion>
			</Card>
		);
		const shownNode = getByText('Shown');
		expect(shownNode).toBeVisible();
		const hiddenNode = getByText('Hidden');
		expect(hiddenNode).not.toBeVisible();
	});

	it('shows expansion when expanded is true', () => {
		const { getByText } = render(
			<Card expanded={true}>
				<p>Shown</p>
				<CardExpansion>Hidden</CardExpansion>
			</Card>
		);
		const shownNode = getByText('Shown');
		expect(shownNode).toBeVisible();
		const hiddenNode = getByText('Hidden');
		expect(hiddenNode).toBeVisible();
	});

	it('expand card toggles expansion state when clicked', () => {
		const { getByText } = render(
			<Card>
				<div>
					<p>Shown</p>
					<ExpandCard>
						<p>Toggle</p>
					</ExpandCard>
				</div>
				<CardExpansion>Hidden</CardExpansion>
			</Card>
		);

		const shownNode = getByText('Shown');
		expect(shownNode).toBeVisible();
		const hiddenNode = getByText('Hidden');
		expect(hiddenNode).not.toBeVisible();

		const toggle = getByText('Toggle');
		toggle.click();

		expect(shownNode).toBeVisible();
		expect(hiddenNode).toBeVisible();
	});

	it('onExpansionChange is triggered when expansion state when clicked', () => {
		const onExpansion = jest.fn();
		const { getByText } = render(
			<Card onExpansionChange={onExpansion}>
				<div>
					<p>Shown</p>
					<ExpandCard>
						<p>Toggle</p>
					</ExpandCard>
				</div>
				<CardExpansion>Hidden</CardExpansion>
			</Card>
		);

		expect(onExpansion).toHaveBeenCalledTimes(0);

		const toggle = getByText('Toggle');
		act(() => toggle.click());

		expect(onExpansion).toHaveBeenCalledTimes(1);

		act(() => toggle.click());

		expect(onExpansion).toHaveBeenCalledTimes(2);
	});

	it('changing forceClose to true closes the expansion', () => {
		let forceClose = false;
		const onExpansion = () => (forceClose = true);
		const { getByText } = render(
			<Card onExpansionChange={onExpansion} expanded={true} forceClose={forceClose}>
				<div>
					<p>Shown</p>
					<ExpandCard>
						<p>Toggle</p>
					</ExpandCard>
				</div>
				<CardExpansion>Hidden</CardExpansion>
			</Card>
		);

		const shownNode = getByText('Shown');
		const hiddenNode = getByText('Hidden');
		expect(shownNode).toBeVisible();
		expect(hiddenNode).toBeVisible();

		const toggle = getByText('Toggle');
		act(() => toggle.click());

		expect(shownNode).toBeVisible();
		expect(hiddenNode).not.toBeVisible();
	});
});
