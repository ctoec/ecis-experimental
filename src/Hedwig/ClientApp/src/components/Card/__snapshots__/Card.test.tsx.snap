// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="oec-card"
  >
    <div
      class="oec-card-cell"
    >
      <p>
        Hi there
      </p>
    </div>
    <div
      class=""
    />
    <div
      class="oec-card-cell oec-card-expansion"
      hidden=""
    />
  </div>
</DocumentFragment>
`;
